File: ocaml/plugins/sigv4.ml
(* AWS Signature Version 4, hand-rolled.

   The AWS providers need exactly one thing from the AWS SDK - request
   signing - and taking the SDK for it would break the no-dependency rule
   that keeps the ports honest. SigV4 is a stable, published algorithm built
   from HMAC-SHA256, which plugins/crypto.ml carries.

   `sigv4` is pure: the caller passes the timestamp, so the same input
   yields the same signature everywhere. That is what lets the shared spec
   carry known-answer cases that all ports must reproduce bit for bit, and
   lets the integration mock recompute the signature server-side. Nothing
   here samples the clock.

   A port of typescript/src/Sigv4.ts, which is canonical. *)

open Secret

(* One request to sign - the same declarative shape the shared spec uses.
   `datetime` is `YYYYMMDDTHHMMSSZ`, and it is the caller's. *)
type signing = {
  smethod : string;
  url : string;
  service : string;
  region : string;
  keyid : string;
  secret : string;
  datetime : string;
  headers : (string * string) list;
  body : string;
  session : string;
}

let nosigning =
  {
    smethod = "";
    url = "";
    service = "";
    region = "";
    keyid = "";
    secret = "";
    datetime = "";
    headers = [];
    body = "";
    session = "";
  }

(* RFC 3986 escaping, which is stricter than the usual URL encoder: AWS
   wants everything but the unreserved set escaped, byte by byte over UTF-8,
   with UPPERCASE hex. `!'()*` are escaped here and are not by the encoders
   most standard libraries offer. *)
let uriescape (text : string) : string =
  let out = Buffer.create (String.length text) in
  String.iter
    (fun ch ->
      let code = Char.code ch in
      if
        ('A' <= ch && ch <= 'Z')
        || ('a' <= ch && ch <= 'z')
        || ('0' <= ch && ch <= '9')
        || '-' = ch || '_' = ch || '.' = ch || '~' = ch
      then Buffer.add_char out ch
      else Buffer.add_string out (Printf.sprintf "%%%02X" code))
    text;
  Buffer.contents out

let hexvalue (ch : char) : int option =
  match ch with
  | '0' .. '9' -> Some (Char.code ch - Char.code '0')
  | 'a' .. 'f' -> Some (Char.code ch - Char.code 'a' + 10)
  | 'A' .. 'F' -> Some (Char.code ch - Char.code 'A' + 10)
  | _ -> None

(* Percent-decode, and nothing else: `+` stays `+`, as on the wire, and a
   malformed escape is kept literally. *)
let uridecode (text : string) : string =
  let out = Buffer.create (String.length text) in
  let len = String.length text in
  let index = ref 0 in

  while !index < len do
    let taken = ref false in

    if '%' = text.[!index] && !index + 2 < len then begin
      match (hexvalue text.[!index + 1], hexvalue text.[!index + 2]) with
      | Some high, Some low ->
        Buffer.add_char out (Char.chr ((high * 16) + low));
        index := !index + 3;
        taken := true
      | _ -> ()
    end;

    if not !taken then begin
      Buffer.add_char out text.[!index];
      incr index
    end
  done;

  Buffer.contents out

(* Split a query into pairs on `&`, then each on its FIRST `=`. A pair with
   no `=` keeps an empty value and is still emitted with one. *)
let canonicalquery (query : string) : string =
  if "" = query then ""
  else
    let pairs =
      List.map
        (fun pair ->
          match String.index_opt pair '=' with
          | None -> (uriescape (uridecode pair), "")
          | Some eq ->
            ( uriescape (uridecode (String.sub pair 0 eq)),
              uriescape (uridecode (String.sub pair (eq + 1) (String.length pair - eq - 1))) ))
        (String.split_on_char '&' query)
    in
    let sorted =
      List.stable_sort
        (fun (lk, lv) (rk, rv) -> if lk = rk then compare lv rv else compare lk rk)
        pairs
    in
    String.concat "&" (List.map (fun (key, value) -> key ^ "=" ^ value) sorted)

(* The url, hand-split into the four parts signing needs.

   Hand-split, not handed to a URL type: twelve URL parsers disagree about
   malformed input, and the signature covers `host` and the raw path, so a
   parser that normalises either of them signs something the server will not
   recompute. *)
type parts = { scheme : string; authority : string; path : string; query : string }

let urlparts (url : string) : parts =
  match findsub url "://" with
  | None -> { scheme = ""; authority = ""; path = url; query = "" }
  | Some mark ->
    let scheme = String.sub url 0 mark in
    let rest = String.sub url (mark + 3) (String.length url - mark - 3) in
    let stop = match splitfirst rest "/?#" with Some at -> at | None -> String.length rest in
    let authority = String.sub rest 0 stop in
    let tail = String.sub rest stop (String.length rest - stop) in
    let path, query =
      match String.index_opt tail '?' with
      | None -> (tail, "")
      | Some at -> (String.sub tail 0 at, String.sub tail (at + 1) (String.length tail - at - 1))
    in
    let path = match String.index_opt path '#' with None -> path | Some at -> String.sub path 0 at in
    { scheme; authority; path = (if "" = path then "/" else path); query }

(* The `host` header as the canonical form wants it: the hostname
   lowercased, any userinfo stripped, and the port appended ONLY when it is
   not the scheme's default - `:443` on https is not what a URL normalises
   to and not what the service recomputes. *)
let hostheader (scheme : string) (authority : string) : string =
  let bare =
    match String.rindex_opt authority '@' with
    | None -> authority
    | Some at -> String.sub authority (at + 1) (String.length authority - at - 1)
  in

  let host, port =
    if String.length bare > 0 && '[' = bare.[0] then
      match String.index_opt bare ']' with
      | None -> (bare, "")
      | Some close ->
        let head = String.sub bare 0 (close + 1) in
        let tail = String.sub bare (close + 1) (String.length bare - close - 1) in
        (head, if String.length tail > 1 && ':' = tail.[0] then String.sub tail 1 (String.length tail - 1) else "")
    else
      match String.rindex_opt bare ':' with
      | None -> (bare, "")
      | Some at ->
        ( String.sub bare 0 at,
          String.sub bare (at + 1) (String.length bare - at - 1) )
  in

  let host = String.lowercase_ascii host in
  let default = if "https" = scheme then "443" else "80" in

  if "" = port || port = default then host else host ^ ":" ^ port

(* Collapse every internal whitespace run - spaces AND tabs - to one space,
   after trimming. AWS folds header values before signing, so `a  b\tc`
   must sign as `a b c` or the service refuses the request. *)
let foldvalue (value : string) : string =
  let trimmed = trim value in
  let out = Buffer.create (String.length trimmed) in
  let inrun = ref false in
  String.iter
    (fun ch ->
      if ' ' = ch || '\t' = ch || '\n' = ch || '\r' = ch then inrun := true
      else begin
        if !inrun then Buffer.add_char out ' ';
        inrun := false;
        Buffer.add_char out ch
      end)
    trimmed;
  Buffer.contents out

(* Sign one request. Answers the headers to attach: authorization,
   x-amz-date, and x-amz-security-token when a session token was given, in
   that order - the spec compares the result as a whole JSON object, so the
   order is contract. *)
let sigv4 (input : signing) : (string * string) list =
  let parts = urlparts input.url in
  let date = String.sub input.datetime 0 (min 8 (String.length input.datetime)) in

  (* The caller's headers first, then host, x-amz-date and the session
     token, so that those three win over anything the caller passed. *)
  let headers = ref [] in
  List.iter
    (fun (key, value) -> headers := assocset !headers (String.lowercase_ascii key) (foldvalue value))
    input.headers;
  headers := assocset !headers "host" (hostheader parts.scheme parts.authority);
  headers := assocset !headers "x-amz-date" input.datetime;
  if "" <> input.session then headers := assocset !headers "x-amz-security-token" input.session;

  let sorted = List.stable_sort (fun (l, _) (r, _) -> compare l r) !headers in

  let canonicalheaders =
    String.concat "" (List.map (fun (key, value) -> key ^ ":" ^ value ^ "\n") sorted)
  in
  let signedheaders = String.concat ";" (List.map fst sorted) in

  let canonicalrequest =
    String.concat "\n"
      [
        String.uppercase_ascii input.smethod;
        parts.path;
        canonicalquery parts.query;
        canonicalheaders;
        signedheaders;
        Crypto.sha256hex input.body;
      ]
  in

  let scope = date ^ "/" ^ input.region ^ "/" ^ input.service ^ "/aws4_request" in

  let stringtosign =
    String.concat "\n"
      [ "AWS4-HMAC-SHA256"; input.datetime; scope; Crypto.sha256hex canonicalrequest ]
  in

  let kdate = Crypto.hmac ("AWS4" ^ input.secret) date in
  let kregion = Crypto.hmac kdate input.region in
  let kservice = Crypto.hmac kregion input.service in
  let ksigning = Crypto.hmac kservice "aws4_request" in
  let signature = Crypto.hex (Crypto.hmac ksigning stringtosign) in

  let out =
    [
      ( "authorization",
        "AWS4-HMAC-SHA256 Credential=" ^ input.keyid ^ "/" ^ scope ^ ", SignedHeaders="
        ^ signedheaders ^ ", Signature=" ^ signature );
      ("x-amz-date", input.datetime);
    ]
  in

  if "" = input.session then out else out @ [ ("x-amz-security-token", input.session) ]

File: haskell/plugins/Sigv4.hs
-- | AWS Signature Version 4, hand-rolled.
--
-- The AWS providers need exactly one thing from the AWS SDK - request
-- signing - and taking the SDK for it would break the no-dependency rule
-- that keeps the ports honest. SigV4 is a stable, published algorithm
-- built from HMAC-SHA256, which "Crypto" supplies in-tree.
--
-- 'sigv4' is pure: the caller passes the timestamp, so the same input
-- yields the same signature everywhere. That is what lets the shared spec
-- carry known-answer cases all ports must reproduce bit for bit, and lets
-- the integration mock recompute the signature server-side. Nothing here
-- samples the clock.
--
-- A port of typescript/plugins/sigv4.ts, which is canonical.

module Sigv4
  ( Signing (..),
    canonicalquery,
    emptysigning,
    sigv4,
    uridecode,
  )
where

import Bytes (hex, utf8decode, utf8encode)
import qualified Data.ByteString as B
import Crypto (hmacsha256, sha256hex)
import Http (uriescape)
import Data.Char (chr, digitToInt, isHexDigit, ord, toLower)
import Data.List (intercalate, isPrefixOf, sortBy)
import Data.Ord (comparing)
import Data.Word (Word8)

-- | One request to sign - the same declarative shape the shared spec uses.
-- @datetime@ is @YYYYMMDDTHHMMSSZ@, and it is the caller's.
data Signing = Signing
  { signmethod :: String,
    signurl :: String,
    signservice :: String,
    signregion :: String,
    signkeyid :: String,
    signsecret :: String,
    signdatetime :: String,
    signheaders :: [(String, String)],
    signbody :: String,
    signsession :: String
  }

emptysigning :: Signing
emptysigning =
  Signing
    { signmethod = "",
      signurl = "",
      signservice = "",
      signregion = "",
      signkeyid = "",
      signsecret = "",
      signdatetime = "",
      signheaders = [],
      signbody = "",
      signsession = ""
    }

-- | Percent-decode, and nothing else: @+@ stays @+@, as on the wire, and
-- a malformed escape is kept literal.
uridecode :: String -> String
uridecode = utf8decode . B.pack . walk
  where
    walk [] = []
    walk ('%' : first : second : rest)
      | isHexDigit first && isHexDigit second =
          (fromIntegral (16 * digitToInt first + digitToInt second) :: Word8) : walk rest
    walk (head : rest) = B.unpack (utf8encode [head]) ++ walk rest

-- | The canonical query string: each half decoded then RFC 3986-escaped,
-- sorted by escaped key then escaped value.
canonicalquery :: String -> String
canonicalquery query
  | null query = ""
  | otherwise = intercalate "&" (map render (sortBy (comparing id) (map pair (spliton '&' query))))
  where
    pair text = case break ('=' ==) text of
      (key, []) -> (uriescape (uridecode key), "")
      (key, _ : value) -> (uriescape (uridecode key), uriescape (uridecode value))

    render (key, value) = key ++ "=" ++ value

spliton :: Char -> String -> [String]
spliton sep body = case break (sep ==) body of
  (piece, []) -> [piece]
  (piece, _ : rest) -> piece : spliton sep rest

-- | The WHATWG @URL.host@: the host lowercased, userinfo stripped, and the
-- port appended only when it is not the scheme's default.
--
-- Hand-split. A platform URL type would have to agree with eleven other
-- ports about malformed input, and none of them do.
authorityof :: String -> String
authorityof url = hostpart ++ portpart
  where
    afterscheme = case break (':' ==) url of
      (_, ':' : '/' : '/' : rest) -> rest
      _ -> url

    scheme = takeWhile (':' /=) url

    authority = takeWhile (\head -> not (elem head "/?#")) afterscheme

    -- Everything after the LAST @, so that userinfo is stripped whether or
    -- not it carries a colon.
    naked = reverse (takeWhile ('@' /=) (reverse authority))

    hostpart = map toLower (if bracketed then take (closing + 1) naked else takeWhile (':' /=) naked)

    bracketed = "[" `isPrefixOf` naked
    closing = maybe (length naked - 1) id (elemindex ']' naked)

    rest' = drop (length hostpart) naked
    port = case rest' of
      (':' : digits) -> digits
      _ -> ""

    portpart
      | null port = ""
      | "https" == scheme && "443" == port = ""
      | "http" == scheme && "80" == port = ""
      | otherwise = ':' : port

elemindex :: Char -> String -> Maybe Int
elemindex want body = go 0 body
  where
    go _ [] = Nothing
    go at (head : rest) = if want == head then Just at else go (at + 1) rest

-- | The path as it appears in the url, already percent-encoded and not
-- re-escaped; @/@ when there is none.
pathof :: String -> String
pathof url = if null raw then "/" else raw
  where
    afterscheme = case break (':' ==) url of
      (_, ':' : '/' : '/' : rest) -> rest
      _ -> url

    raw = takeWhile (\head -> not (elem head "?#")) (dropWhile (\head -> '/' /= head) afterscheme)

queryof :: String -> String
queryof url = case break ('?' ==) url of
  (_, []) -> ""
  (_, _ : rest) -> takeWhile ('#' /=) rest

-- | Trim, then collapse every internal run of spaces and tabs to one.
-- AWS folds sequential whitespace before signing, so @a  b\\tc@ must sign
-- as @a b c@ or the service refuses the request.
foldspace :: String -> String
foldspace = unwords . words

-- | Sign one request. Answers the headers to attach: @authorization@,
-- @x-amz-date@, and @x-amz-security-token@ only when a session was given,
-- in that order - the spec compares the whole map, so the order is
-- contract.
sigv4 :: Signing -> [(String, String)]
sigv4 input =
  [("authorization", authorization)]
    ++ [("x-amz-date", signdatetime input)]
    ++ [("x-amz-security-token", signsession input) | not (null (signsession input))]
  where
    date = take 8 (signdatetime input)

    -- The caller's headers first, then host, x-amz-date and the session
    -- token, so that those win over anything the caller passed.
    given = [(map toLower key, foldspace value) | (key, value) <- signheaders input]

    withhost =
      foldl set given $
        [("host", authorityof (signurl input)), ("x-amz-date", signdatetime input)]
          ++ [("x-amz-security-token", signsession input) | not (null (signsession input))]

    set entries (key, value) =
      if any ((key ==) . fst) entries
        then map (\(at, was) -> if key == at then (at, value) else (at, was)) entries
        else entries ++ [(key, value)]

    headers = sortBy (comparing fst) withhost

    canonicalheaders = concat [key ++ ":" ++ value ++ "\n" | (key, value) <- headers]
    signedheaders = intercalate ";" (map fst headers)

    canonicalrequest =
      intercalate
        "\n"
        [ map upperhead (signmethod input),
          pathof (signurl input),
          canonicalquery (queryof (signurl input)),
          canonicalheaders,
          signedheaders,
          sha256hex (signbody input)
        ]

    scope = date ++ "/" ++ signregion input ++ "/" ++ signservice input ++ "/aws4_request"

    stringtosign =
      intercalate
        "\n"
        ["AWS4-HMAC-SHA256", signdatetime input, scope, sha256hex canonicalrequest]

    kdate = hmacsha256 (utf8encode ("AWS4" ++ signsecret input)) (utf8encode date)
    kregion = hmacsha256 kdate (utf8encode (signregion input))
    kservice = hmacsha256 kregion (utf8encode (signservice input))
    ksigning = hmacsha256 kservice (utf8encode "aws4_request")
    signature = hex (hmacsha256 ksigning (utf8encode stringtosign))

    authorization =
      "AWS4-HMAC-SHA256 Credential="
        ++ signkeyid input
        ++ "/"
        ++ scope
        ++ ", SignedHeaders="
        ++ signedheaders
        ++ ", Signature="
        ++ signature

    upperhead head = if 'a' <= head && 'z' >= head then chr (ord head - 32) else head
